includes/orders/functions/refunds.php

```diff
@@ -307,264 +308,110 @@ function edd_refund_order( $order_id = 0, $order_items = array() ) {
-				'object_id'   => $order_item_id,
-				'object_type' => 'order_item',
-				'type_id'     => $adjustment->type_id,
-				'type_key'    => $adjustment->type_key,
-				'type'        => $adjustment->type,
-				'description' => $adjustment->description,
-				'subtotal'    => edd_negate_amount( $adjustment->subtotal ),
-				'tax'         => edd_negate_amount( $adjustment->tax ),
-				'total'       => edd_negate_amount( $adjustment->total ),
-			) );
+		$new_item_id = edd_add_order_item( $order_item );
+
+		if ( ! empty( $order_item['parent'] ) ) {
+			$order_item_id_map[ $order_item['parent'] ] = $new_item_id;
 		}
 
-		// Set the order item as 'refunded' on the original order.
-		edd_update_order_item( $item->id, array( 'status' => 'refunded' ) );
+		// Update the status on the original order item.
+		if ( ! empty( $order_item['parent'] ) && ! empty( $order_item['original_item_status'] ) ) {
+			edd_update_order_item( $order_item['parent'], array( 'status' => $order_item['original_item_status'] ) );
+		}
 	}
 
 	/** Insert order adjustments **********************************************/
 
-	foreach ( $order->adjustments as $adjustment ) {
-		edd_add_order_adjustment( array(
-			'object_id'   => $new_order_id,
-			'object_type' => 'order',
-			'type_id'     => $adjustment->type_id,
-			'type_key'    => $adjustment->type_key,
-			'type'        => $adjustment->type,
-			'description' => $adjustment->description,
-			'subtotal'    => edd_negate_amount( $adjustment->subtotal ),
-			'tax'         => edd_negate_amount( $adjustment->tax ),
-			'total'       => edd_negate_amount( $adjustment->total ),
-		) );
+	foreach( $validator->get_refunded_adjustments() as $adjustment ) {
+		if ( ! empty( $adjustment['object_type'] ) && 'order' === $adjustment['object_type'] ) {
+			$adjustment['object_id'] = $refund_id;
+		} elseif ( ! empty( $adjustment['object_type'] ) && 'order_item' === $adjustment['object_type'] ) {
+			/*
+			 * At this point, `object_id` references an order item which is attached to the
+			 * original order record. We need to try to convert this to a _refund_ order item
+			 * instead.
+			 *
+			 * If we can't (such as, if the order item was never refunded), we'll have to
+			 * convert the adjustment to be an `order` object type instead. That's because we
+			 * _have_ to reference a refund object of some kind.
+			 */
+			$should_convert = true;
+			if ( ! empty( $adjustment['object_id'] ) ) {
+				/*
+				 * First check in our map, as it avoids DB queries.
+				 * We should hit this if the order item was refunded in this same transaction.
+				 */
+				if ( ! empty( $order_item_id_map[ $adjustment['object_id'] ] ) ) {
+					$adjustment['object_id'] = $order_item_id_map[ $adjustment['object_id'] ];
+					$should_convert   = false;
+				} else {
+					// Otherwise we'll have to do a DB query, which would pick up previously refunded items.
+					$refund_order_item = edd_get_order_item_by( 'parent', $adjustment['object_id'] );
```

Comment:
I left the comment about the refunded fee not showing properly on a refund and I think, as I read it, that this logic is where it's gotten snarled, because I had already done a partial refund on the related `order_item`. However, in the refund I was creating, I was not refunding the item at all, but I think this logic will have returned the order item anyway.

This may be a case of I was doing it wrong, but I managed it. Can this be accounted for?

Reply:
Update: I refunded the order item fee, but not the order item, on an order which had no refunds applied, and the fee shows properly on the refunds screen.

Reply:
Just going to remove that `else` statement completely. Think that will sort it.